### PR TITLE
Side average/integral for material properties

### LINKED ISSUE #18649: Side average reporters for material properties
## Reason
Create reporters to compute averages of material properties along side sets (and possibly also over blocks).

## Design
Add reporters templated on the material property type that is to be averaged.

## Impact
Added capability (as far as I can see we have no SideAverageMaterialProperty PP, and since Reporters are the hot $#!+ these days, we might as well make it general and permit averaging of vectors and tensors as well).

### PR DESCRIPTION
**Pull request by dschwen**
Closes #18649

There are two new concepts in this PR to avoid code duplication in classes that perform some operation on a component of an indexable material property (Real, RealVectorValue, RankTwoTensor, RankThreeTensor, RankFourTensor, etc.). 

- One is the `ComponentUtils<T>` template which allows handles all the addition of index parameters and value extraction from the property. With that it is super easy to instantiate a zillion MOOSE objects for the various material property types. See 739f21e7248a2171e52f527eaf02dd67a39973fa for how this looks. 
-Two is using a proxy object to get and evaluate a material property that could be of any of the above types. With this tool you can write an object that will automatically deal correctly with the property type that is passed in. The upside is that we are only registering one object (rather than five+) for a certain functionality like side integral). See [here](https://github.com/idaholab/moose/pull/18661/files#diff-f6f5fe5b83c6a6013f5aff19ad813be96d0cc01aef8ef1d985a0c282788ffee3) and [here](https://github.com/idaholab/moose/pull/18661/files#diff-b8b57752d4c10853a4310bdf59cc654b87fbcddf519a6118acb01e486afc74cf) for how this looks (I really like how `_prop[_qp]` will always just return the component of `mat_prop` indexed by `component`).

I think I prefer the second method as it doesn't clutter up the list of MOOSE objects with similar stuff as much.

P.S.: The `IndexableProperty` can be evaluated as `bool` to check if any matching property was found. That way a hybrid AD/non-AD class could even be built.

```
  const IndexableProperty<SideIntegralPostprocessor, false> _prop;
  const IndexableProperty<SideIntegralPostprocessor, true> _ad_prop;
```
and 
```
if (_prop)
  return _prop[_qp];
if(_ad_prop)
  return MetaPhysicL::raw_value(_ad_prop[_qp]);
mooseError("Neither AD nor non-AD property found");
```

### CODE REVIEW

File: framework/include/utils/IndexableProperty.h
Comment (bwspenc) [inline framework/include/utils/IndexableProperty.h:53]: Do we have a convention for the parameter name for a material property? It seems like `property` or something would be more consistent with other classes, but I'd need to look around for examples.

File: framework/include/utils/IndexableProperty.h
Comment (bwspenc) [inline framework/include/utils/IndexableProperty.h:77]: My first thought when I saw this is that we need to check to make sure we don't have more than one of these properties, but I guess MOOSE won't allow us to have properties of different types with the same name, right?

File: framework/include/utils/IndexableProperty.h
Comment (bwspenc) [inline framework/include/utils/IndexableProperty.h:94]: It seems like this error should also be in the constructor so it's hit right away. If you did that, it wouldn't really be necessary here.

**Review by bwspenc — DISMISSED** [review]
I like the design using `IndexableProperty`.  Should `ComponentUtils` be deleted, or do you anticipate that being used for something else? This all needs testing and documentation.

File: framework/include/utils/IndexableProperty.h
Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:53]: We use add param  `mat_prop` 12 times and `property` 4 times

Comment (bwspenc) [reply framework/include/utils/IndexableProperty.h:53]: Bummer. I like `property` better. It's actually shorter, and I'm in favor of avoiding abbreviations when possible.

Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:54]: `mat_prop`
`property`

same length. I agree regarding abbreviations. I don't think I came up with `mat_prop`. If you want me to change it, weighing consistency vs. nicer parameter name, I'd be happy to do it.

File: framework/include/utils/IndexableProperty.h
Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:79]: Correct!

File: framework/include/utils/IndexableProperty.h
Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:98]: That would add more code for the check (both written _and_ executed).

File: framework/doc/content/source/postprocessors/SideAverageMaterialProperty.md
Comment (GiudGiud) [inline framework/doc/content/source/postprocessors/SideAverageMaterialProperty.md:7]:
```suggestion
supplied to select a scalar component from the supplied material property (e.g.
```

File: framework/doc/content/source/postprocessors/SideIntegralMaterialProperty.md
Comment (GiudGiud) [inline framework/doc/content/source/postprocessors/SideIntegralMaterialProperty.md:1]:
```suggestion
# SideIntegralMaterialProperty
```

File: framework/doc/content/source/postprocessors/SideIntegralMaterialProperty.md
Comment (GiudGiud) [inline framework/doc/content/source/postprocessors/SideIntegralMaterialProperty.md:14]:
```suggestion
!syntax parameters /Postprocessors/SideIntegralMaterialProperty

!syntax inputs /Postprocessors/SideIntegralMaterialProperty

!syntax children /Postprocessors/SideIntegralMaterialProperty
```

File: framework/doc/content/source/postprocessors/SideIntegralMaterialProperty.md
Comment (GiudGiud) [inline framework/doc/content/source/postprocessors/SideIntegralMaterialProperty.md:7]:
```suggestion
supplied to select a scalar component from the supplied material property (e.g.
```

File: test/tests/postprocessors/side_material.i
Comment (GiudGiud) [inline test/tests/postprocessors/side_material.i:45]: should you grab GenericConstantArray and GenericConstant2DArray (and support the RealEigenVector (indexable?))

File: test/tests/postprocessors/side_integral/tests
Comment (GiudGiud) [inline test/tests/postprocessors/side_integral/tests:57]:
```suggestion
      detail = 'and shall error out if the index components do not match the dimension of the '
```

File: test/tests/postprocessors/side_integral/tests
Comment (GiudGiud) [inline test/tests/postprocessors/side_integral/tests:41]: we usually have the requirements try to be code agnostic, so like 'of vector type' or 'for vector material properties'

File: test/tests/postprocessors/side_average_value/tests
Comment (GiudGiud) [inline test/tests/postprocessors/side_average_value/tests:45]:
```suggestion
      detail = 'And shall error out if the index components do not match the dimension of the '
```

File: test/tests/postprocessors/side_material.i
Comment (GiudGiud) [reply test/tests/postprocessors/side_material.i:45]: do you know who uses those actually

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [inline framework/include/utils/IndexableProperty.h:40]: does it need the host only for starting with the right valid params? Couldn't we keep that part in the host validParams?
I feel like this could be a lot less instantiations of the template

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [inline framework/include/utils/IndexableProperty.h:15]: docstring

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [inline framework/include/utils/IndexableProperty.h:46]: docstrings

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [reply framework/include/utils/IndexableProperty.h:98]: still better to error at construction time when possible imo

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: framework/include/utils/IndexableProperty.h
Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:40]: Needed also for getParam and getMateriaProperty below

File: test/tests/postprocessors/side_material.i
Comment (dschwen) [reply test/tests/postprocessors/side_material.i:45]: I don't. Adding those later -  if required - would be another possibility.

**Comment by dschwen** [comment]
> still better to error at construction time when possible imo

I can neither reply to that comment, nor make this as resolved. what's going on here?

image

**Comment by GiudGiud** [comment]
it s because it s based on outdated code. But if you look at the number of unresolved conversation you have to resolve, 3, it only includes the one you could resolve right now, not the outdated ones.

**Comment by GiudGiud** [comment]
but it goes to show we cant rely on PR comments on code at all, since github can make them disappear. We have to keep the conversations on the PR or issue main threads

**Comment by dschwen** [comment]
Really, because here's one based on outdated code that I can comment on and resolve:

image

(and the one mentioned above I could neither comment on or resolve right after you added it)

**Comment by dschwen** [comment]
I think GitHub just glitched for me. The unresolvable comment is resolved now.

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [inline framework/include/utils/IndexableProperty.h:86]: if you passed params instead of host couldn't you remove host?

**Review by GiudGiud — DISMISSED** [review]

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [inline framework/include/utils/IndexableProperty.h:138]: does it make more sense to have the material property interface at the host level or at the IndexableProperty? I feel like the first one is more convoluted but probably more convenient?

Comment (GiudGiud) [reply framework/include/utils/IndexableProperty.h:138]: the second one doesnt work I think.

I think what I would like best would be to see this code work within the MaterialPropertyInterface. Maybe there's a hold up there I have not thought things through as deeply

File: framework/include/utils/IndexableProperty.h
Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:86]: I had replied to that already

> Needed also for getParam and **getMateriaProperty** below

File: framework/include/utils/IndexableProperty.h
Comment (dschwen) [reply framework/include/utils/IndexableProperty.h:138]: let's merge this as is now and if it turns out to be of greater interest we can always migrate it into the mat prop interface

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [reply framework/include/utils/IndexableProperty.h:86]: true true. I was working my way to my last comment ^^
getMaterialProp comes along with the MatPropInterface

File: framework/include/utils/IndexableProperty.h
Comment (GiudGiud) [reply framework/include/utils/IndexableProperty.h:138]: I'm ok with that if you need this object now.  But I'm not a fan of this interface and if we deprecate it soon then should we add a note to the documentation or source to advise developers to contact us if they want to use this?
What has been done for less ideal merges before?

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**
